pyiceberg/table/update/validate.py

```diff
@@ -0,0 +1,72 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Optional
+
+from pyiceberg.manifest import ManifestContent, ManifestFile
+from pyiceberg.table import Table
+from pyiceberg.table.snapshots import Operation, Snapshot, ancestors_between
+
+
+class ValidationException(Exception):
+    """Raised when validation fails."""
+
+
+def validation_history(
+    table: Table,
+    from_snapshot: Snapshot,
+    to_snapshot: Optional[Snapshot],
```

Comment:
What do you think of making the API a bit more strict and simplify code along the way. I don't think this needs to be `Optional`:
```suggestion
    to_snapshot: Snapshot,
```

Reply:
done

Reply:
@Fokko I think the existing behavior of the Spark/Java API is for to check all snapshots if the `from_snapshot` is Optional.

https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java#L443-L448

Should we mirror that behavior in PyIceberg, or just require that the `validate_from_snapshot_id` be specified if `isolation_level` is not None?

Reply:
My initial thought was that we could always widen the API and make it optional.

Looking at it more closely, only the `SparkWriteConf` is optional, we only use the value if is `!= null`:

- https://github.com/apache/iceberg/blob/d7f8e5400519f84cd2af82d7769713b24e916809/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java#L317-L321
- https://github.com/apache/iceberg/blob/d7f8e5400519f84cd2af82d7769713b24e916809/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java#L361-L366